### ISSUE TITLE
Helix SharpDX is not picking the best device.

### ISSUE THREAD

**Issue by ikeough** [helix-toolkit/helix-toolkit#153]
On systems which have Intel (R) HD Graphics alongside another, more powerful, gpu, the `Effects.GetBestAdapter()` method is not actually returning the "best" adapter. This check is based solely on the feature level. If the second adapter in the list returned by `Factory.Adapters` has the same feature level it will not be set as the "best" adapter because the comparison uses `>`. 

There needs to be a better heuristic for selecting the "best" display adapter. It could be a combination of the feature level and the available memory. I'm not sure, but I'm pretty sure on systems that have Intel (R) HD cards alongside NVIDIA cards (ex. Dell Precision 4700), that the Intel card is being used by default. I'll do more research and follow up.

UPDATE:
I had a chance to debug on a machine with an Intel Graphics setup. It turns out that those systems are not reporting two separate display adapters, but one...the Intel. So making a better switch condition is not an option. Interestingly, the Intel cards report feature level 11 support. But, when you call `CheckMultisampleQualityLevels` with any sample count and any `Format`, you get either 0 or -1 for `sampleQuality`. This is in contrast to my NVIDIA card where `CheckMultisampleQualityLevels` with a sample count of 8 returns 32 for `sampleQuality`. Helix falls back to using a `sampleCount` and `qualityLevel` of 0 and 1 respectively, effectively disabling MSAA. 

Just to be clear, on any system with an Intel Graphics card, Helix does NOT antialias. This is my first rodeo with DirectX, so I'm looking for anyone with more experience to suggest how we get antialiasing to work with these cards, especially because their feature level _should_ support it.

**Comment by ikeough** [comment]
I've had an opportunity to test on a system with multiple adapters, an Intel Graphics 4600 and an NVIDIA . They both show the same feature level, 11, but the logic for picking an adapter defaults to the Intel because it's first in the list. Using a further comparison for which adapter has the most dedicated video memory get it to use the NVIDIA (the Intel has 0 dedicated video memory). I can't confirm that it's working yet because I'm debugging this on a remote system and I've run into a further problem which is that the OnRenderSizeChanged method is not being called on the remote machine when the window size is changed. As such, it's not calling `CreateAndBindTargets` after the initial load and the `DPFCanvas` has an `ActualHeight` and `ActualWidth` of 0, so Helix defaults to setting them both to 100. Calling the constructor for `Texture with a width and height equal to 100 into ` throws a `SharpDX` exception with an "incorrect parameter" warning. Still working on this.

**Comment by JeremyAnsel** [comment]
The actual code of `GetBestAdapter` returns the first adapter with the highest feature level.
On systems with an hardware adapter that supports feature level 10, `GetBestAdapter` will choose the WARP device (feature level 11).

I agree that a better heuristic is needed. And maybe let the user choose which adapter to use.

**Comment by ikeough** [comment]
@JeremyAnsel Thanks! In [a separate issue](https://github.com/helix-toolkit/helix-toolkit/issues/161) I point out that the WARP driver can't be specified at all (I don't think). That is, if instead of using `GetBestAdapter`, I just tell it to use the WARP driver `factory.Adapters[1]`(on my machine) it just doesn't work. If fails to create a texture when attempting to set the `renderTarget` on `DX11ImageSource`. We get a blank screen in the `Simple Demo` application.

Have you been able to set WARP through code and have it work? I'm on Windows 8.1 so perhaps that is the issue.

**Comment by BetaRavener** [comment]
I'm currently also trying to use dedicated mobile graphics card with Helix-Toolkit. I have notebook which has Intel HD 4000 and NVidia 635M in Optimus setup, Windows 10 system.

What I experience is that the adapter would be selected properly - the NVidia has the same features as Intel but more memory, so it is better adapter. However, there's a "fix":
``` 
// Windows 10 fix
if (item.Outputs == null || item.Outputs.Length == 0)
{
  continue;
}
```
The NVidia graphics fails to pass this and so it gets discarded. This is kind of proper action, because skipping this check produces blank screen. I have also tried to select NVidia GPU manually as Accelerator in NVidia Control Panel, but it still has no Outputs.

However, this should get fixed properly, not just by discarding the GPU. From Wikipedia: "When a user launches an application, the graphics driver tries to determine whether the application would benefit from the discrete GPU. If so, the GPU is powered up from an idle state and is passed all rendering calls. Even in this case, though, the integrated graphics processor (IGP) is used to output the final image". 

This means that the NVidia really doesn't have any Outputs but still should be used for rendering, and then the final image should be just transferred to Intel graphics for displaying. This is problem for every user with Optimus in their system, so I think implementing this properly would benefit a lot of people.

**Comment by JeremyAnsel** [comment]
This was fixed in #164, #228 and #508.

**Comment by holance** [comment]
Fixed
